### PR TITLE
Implement lookupResolvedPackageUris and lookupPackageUris

### PR DESCRIPTION
**Pull request by annagrin**
- implement the new API from vm service 7.4.0
- add tests
- does not suport dart: scheme uris yet:
  https://github.com/dart-lang/webdev/issues/1457

Closes: https://github.com/dart-lang/webdev/issues/1446

### CODE REVIEW

File: dwds/lib/src/utilities/dart_uri.dart [hunk 1]
```diff
@@ -51,38 +51,47 @@ class DartUri {
   static PackageConfig _packageConfig;
 
   /// All of the known libraries, indexed by their absolute file URL.
-  static final Map<String, String> _libraryNamesByPath = {};
+  static final Map<String, String> _resolvedUriToUri = {};
+
+  /// All of the known absolute library paths, indexed by their library URL.
```
Comment (nshahan) [inline dwds/lib/src/utilities/dart_uri.dart:56]: Can you give examples of the mapping in these two comments. I'm having trouble imagining what all of these strings would actually look like.

Also, depending on what these strings look like and where they come from, do we need to worry about cases where the same library is imported as different URIs? For example a relative uri and a package uri?

`import 'package:foo/src/bar.dart';` vs `import '../src/dart';`

**Review by nshahan — APPROVED** [review]
LGTM with one comment.

File: dwds/lib/src/utilities/dart_uri.dart (hunk 1 above)
Comment (annagrin) [reply dwds/lib/src/utilities/dart_uri.dart:56]: > Also, depending on what these strings look like and where they come from, do we need to worry about cases where the same library is imported as different URIs? For example a relative uri and a package uri?
> `import 'package:foo/src/bar.dart'; vs import '../src/dart';`

I am just assuming that all library uris (coming from https://github.com/dart-lang/sdk/blob/main/runtime/vm/service/service.md#getscripts) can be translated to their absolute paths and back.

I only see library uris in forms of 
`dart:html`
`org-dartlang-app:///example/hello_world/main.dart`
`package:path/path.dart`,
`package:path/src/path_set.dart`

Anything imported with a relative path shows as a `org-dartlang-app` path in the library uri so I assume we don't have to care about relative paths.

Comment (annagrin) [reply dwds/lib/src/utilities/dart_uri.dart:56]: Added examples.

File: dwds/lib/src/utilities/dart_uri.dart [hunk 2]
```diff
@@ -50,39 +50,70 @@ class DartUri {
...
   /// indexed by absolute file: URI.
   static void _recordAbsoluteUri(String libraryUri) {
     var uri = Uri.parse(libraryUri);
-    if (uri.scheme == 'dart' ||
-        (uri.scheme == '' && !uri.path.endsWith('.dart'))) {
-      // We ignore dart: libraries, and non-Dart libraries referenced by path.
-      // e.g. main.dart.bootstrap
-      // TODO(alanknight): These should not be showing up in the library list,
-      // fix _getLibraryRefs and then remove this check.
+    if (uri.scheme == '' && !uri.path.endsWith('.dart')) {
```
Comment (bkonyi) [inline dwds/lib/src/utilities/dart_uri.dart:101]: Nit: should this be `uri.scheme.isEmpty`?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 3]
```diff
@@ -50,39 +50,70 @@ class DartUri {
...
       // separators, so we can join them as url paths to get the absolute file
       // url.
       var libraryPath = p.url.join(currentDirectoryUri, uri.path.substring(1));
-      _libraryNamesByPath[libraryPath] = libraryUri;
+      _uriToResolvedUri[libraryUri] = libraryPath;
+      _resolvedUriToUri[libraryPath] = libraryUri;
     } else if (uri.scheme == 'package') {
       var libraryPath = _packageConfig.resolve(uri);
-      _libraryNamesByPath['$libraryPath'] = libraryUri;
+      _uriToResolvedUri[libraryUri] = '$libraryPath';
```
Comment (bkonyi) [inline dwds/lib/src/utilities/dart_uri.dart:115]: Nit: can't we just do `libraryPath.toString()`?

File: dwds/test/chrome_proxy_service_test.dart [hunk 4]
```diff
@@ -1261,6 +1261,99 @@ void main() {
           service.getRetainingPath(null, null, null), throwsRPCError);
     });
 
+    test('lookupResolvedPackageUris converts package and org-dartlang-app uris',
+        () async {
+      var vm = await service.getVM();
+      var isolateId = vm.isolates.first.id;
+      var scriptList = await service.getScripts(isolateId);
+
+      var uris = [...scriptList.scripts.map((e) => e.uri)];
```
Comment (bkonyi) [inline dwds/test/chrome_proxy_service_test.dart:1270]: Nit: `scriptList.scripts.map((e) => e.uri).toList()`?

File: dwds/test/chrome_proxy_service_test.dart [hunk 5]
```diff
@@ -1261,6 +1261,99 @@ void main() {
...
+      ]);
+    }, skip: 'https://github.com/dart-lang/webdev/issues/1457');
+
+    test('lookupPackageUris finds package and org-dartlang-app paths',
+        () async {
+      var vm = await service.getVM();
+      var isolateId = vm.isolates.first.id;
+      var scriptList = await service.getScripts(isolateId);
+
+      var uris = [...scriptList.scripts.map((e) => e.uri)];
```
Comment (bkonyi) [inline dwds/test/chrome_proxy_service_test.dart:1316]: Nit: same as above

File: dwds/test/chrome_proxy_service_test.dart [hunk 6]
```diff
@@ -1261,6 +1261,99 @@ void main() {
...
+      expect(
+          resolvedUris.uris,
+          containsAll([
+            contains('/_test/example/hello_world/main.dart'),
+            contains('/lib/path.dart'),
+            contains('/lib/src/path_set.dart'),
+          ]));
+    });
+
+    test('lookupResolvedPackageUris does not translate non-exsitent paths',
```
Comment (bkonyi) [inline dwds/test/chrome_proxy_service_test.dart:1283]: Nit: "non-exsitent"->"non-existent" here and below

**Review by bkonyi — APPROVED** [review]
LGTM!

File: dwds/lib/src/utilities/dart_uri.dart (hunk 2 above)
Comment (annagrin) [reply dwds/lib/src/utilities/dart_uri.dart]: Done!

File: dwds/lib/src/utilities/dart_uri.dart (hunk 3 above)
Comment (annagrin) [reply dwds/lib/src/utilities/dart_uri.dart]: Done!

File: dwds/test/chrome_proxy_service_test.dart (hunk 5 above)
Comment (annagrin) [reply dwds/test/chrome_proxy_service_test.dart]: Done!

File: dwds/test/chrome_proxy_service_test.dart (hunk 6 above)
Comment (annagrin) [reply dwds/test/chrome_proxy_service_test.dart]: done, thanks!

**Comment by annagrin** [comment]
@bkonyi I actually figured out how to resolve SDK paths, added it to this PR. Appreciate another review!

**Review by bkonyi — COMMENTED** [review]
Updated changes LGTM (sorry for the delay, was OOO)

File: dwds/pubspec.yaml [hunk 7]
```diff
@@ -10,6 +10,7 @@ environment:
   sdk: ">=2.15.0 <3.0.0"
 
 dependencies:
+  _fe_analyzer_shared: ^31.0.0
```
Comment (natebosch) [inline dwds/pubspec.yaml:13]: This dependency is very unfortunate. Do we have a plan to migrate away from it?

cc @scheglov

Comment (annagrin) [reply dwds/pubspec.yaml:13]: Yes, it would be part of https://github.com/dart-lang/webdev/issues/1290

Comment (annagrin) [reply dwds/pubspec.yaml:13]: Required design and includes work in DDC, webdev, build, flutter tools, DDR, Bolt, ...

File: dwds/lib/src/utilities/dart_uri.dart [hunk 8]
```diff
@@ -158,4 +84,161 @@ class DartUri {
...
+  /// from the server. For example, 'hello_world/main.dart' or
+  /// 'packages/path/src/utils.dart'.
+  final String serverPath;
+
+  static final _logger = Logger('DartUri');
+
+  /// The way we resolve file: URLs into package: URLs
+  static PackageConfig _packageConfig;
+
+  /// The way we resolve dart: URLs into org-dartland-sdk: URLs
```
Comment (nshahan) [inline dwds/lib/src/utilities/dart_uri.dart:100]: dartland -> dartlang?

File: dwds/test/chrome_proxy_service_test.dart [hunk 9]
```diff
@@ -1295,6 +1295,101 @@ void main() {
...
+      var resolvedUris =
+          await service.lookupResolvedPackageUris(isolateId, uris);
+
+      expect(
+          resolvedUris.uris,
+          containsAll([
+            contains('/_test/example/hello_world/main.dart'),
+            contains('/lib/path.dart'),
+            contains('/lib/src/path_set.dart'),
+          ]));
```
Comment (nshahan) [inline dwds/test/chrome_proxy_service_test.dart:1314]: I'm a little confused why expectation needs `containsAll()` with nested `contains()`. Is that on purpose?

**Review by nshahan — COMMENTED** [review]
LGTM

File: dwds/test/chrome_proxy_service_test.dart (hunk 9 above)
Comment (annagrin) [reply dwds/test/chrome_proxy_service_test.dart:1314]: containsAll matches the array, and the nested contains match the elements. I like that our matches allow that!

File: dwds/lib/src/utilities/dart_uri.dart (hunk 8 above)
Comment (annagrin) [reply dwds/lib/src/utilities/dart_uri.dart:100]: thanks!

File: dwds/lib/src/services/expression_compiler_service.dart [hunk 10]
```diff
@@ -277,16 +273,14 @@ class ExpressionCompilerService implements ExpressionCompiler {
...
-    var sdkRoot = _sdkRoot ?? p.join(sdkDir, 'lib', '_internal');
+    var sdkSummaryRoot = p.join(sdkDir, 'lib', '_internal');
     var sdkSummaryPath = soundNullSafety
-        ? p.join(sdkRoot, 'ddc_outline_sound.dill')
-        : p.join(sdkRoot, 'ddc_sdk.dill');
-    var librariesPath =
-        _librariesPath ?? p.join(sdkDir, 'lib', 'libraries.json');
+        ? p.join(sdkSummaryRoot, 'ddc_outline_sound.dill')
+        : p.join(sdkSummaryRoot, 'ddc_sdk.dill');
+    var librariesPath = p.join(sdkDir, 'lib', 'libraries.json');
```
[the remark this answers is not in the document]
Comment (annagrin) [reply dwds/lib/src/services/expression_compiler_service.dart:283]: this is much later than the dart_uri calls. Will investigate if we can postpone dart uri registration until loading the bootstrap and reading the metadata.

**Pull request merged**